packages/docusaurus-mdx-loader/src/remark/toc/search.ts

```diff
@@ -8,40 +8,75 @@
 import toString from 'mdast-util-to-string';
 import visit, {Visitor} from 'unist-util-visit';
 import {toValue} from '../utils';
-import type {TOCItem as TOC} from '@docusaurus/types';
+import type {TOCItem} from '@docusaurus/types';
 import type {Node} from 'unist';
 import type {Heading} from 'mdast';
 
-// Visit all headings. We `slug` all headings (to account for
-// duplicates), but only take h2 and h3 headings.
-export default function search(node: Node): TOC[] {
-  const headings: TOC[] = [];
-  let current = -1;
-  let currentDepth = 0;
+// Intermediate interface for TOC algorithm
+interface SearchItem {
+  node: TOCItem;
+  level: number;
+  parentIndex: number;
+}
+
+/**
+ *
+ * Generate a TOC AST from the raw Markdown contents
+ */
+export default function search(node: Node): TOCItem[] {
+  const headings: SearchItem[] = [];
 
   const visitor: Visitor<Heading> = (child, _index, parent) => {
     const value = toString(child);
 
-    if (parent !== node || !value || child.depth > 3 || child.depth < 2) {
+    // depth:1 headings are titles and not included in the TOC
+    if (parent !== node || !value || child.depth < 2) {
       return;
     }
 
-    const entry: TOC = {
-      value: toValue(child),
-      id: child.data!.id as string,
-      children: [],
-    };
-
-    if (!headings.length || currentDepth >= child.depth) {
-      headings.push(entry);
-      current += 1;
-      currentDepth = child.depth;
-    } else {
-      headings[current].children.push(entry);
-    }
+    headings.push({
+      node: {
+        value: toValue(child),
+        id: child.data!.id as string,
+        children: [],
+        level: child.depth,
+      },
+      level: child.depth,
+      parentIndex: -1,
+    });
   };
 
   visit(node, 'heading', visitor);
 
-  return headings;
+  // Keep track of which previous index would be the current heading's direcy parent.
+  // Each entry <i> is the last index of the `headings` array at heading level <i>.
+  // We will modify these indices as we iterate through all headings.
+  // e.g. if an ### H3 was last seen at index 2, then prevIndexForLevel[3] === 2
+  // indices 0 and 1 will remain unused.
+  const prevIndexForLevel = Array(7).fill(-1);
+
+  headings.forEach((curr, currIndex) => {
+    // take the last seen index for each ancestor level. the highest
+    // index will be the direct ancestor of the current heading.
+    const ancestorLevelIndexes = prevIndexForLevel.slice(2, curr.level);
```

Comment:
`Math.max(...prevIndexForLevel.slice(2, 2)) = Math.max(...[]) = -Infinity`, although that doesn't break the algorithm, does it really make sense?

Reply:
🤪 my head hurts 🤯 will have to take more time reviewing this.

We need more test cases to cover the edge cases

Reply:
I tested a few edge cases locally (never committed the changes) and it _seemed_ to work somewhat well.

My personal take (as a drive-by contributor, feel free to ignore) is that while the TOC should _never break_, documents that have wonky heading formatting won't suffer too much from having a wonky TOC.

Reply:
Agree, just added tests to ensure the behavior with wonky TOC stays the same over time :)